Entities/AHuman.cpp

```diff
@@ -1796,6 +1803,39 @@ bool AHuman::UpdateMovePath()
     return true;
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+
+void AHuman::UpdateWalkAngle(AHuman::Layer whichLayer) {
+	if (m_Controller.IsState(BODY_JUMP)) {
+		m_WalkAngle[whichLayer] = Matrix(c_QuarterPI * GetFlipFactor());
+	} else {
+		// Cast rays to calculate the approximate shape of terrain.
+		int rayCount = 4;
+		float rayLength = 10.0F;
+		Vector hipPos = m_Pos;
+		if (whichLayer == AHuman::Layer::FGROUND && m_pFGLeg) {
+			rayLength += m_pFGLeg->GetMaxLength();
+			hipPos += RotateOffset(m_pFGLeg->GetParentOffset());
+		} else if (m_pBGLeg) {
+			rayLength += m_pBGLeg->GetMaxLength();
+			hipPos += RotateOffset(m_pBGLeg->GetParentOffset());
+		}
+		float traceRotation = rayCount > 1 ? c_HalfPI / static_cast<float>(rayCount - 1) * GetFlipFactor() : 0;
+		Vector hitPos;
+		Vector terrainVector(0, rayLength);
```

Comment:
Why do you start terrainVector 10 (or whatever) pixels below? Just to try and enforce forwards walkangle (i.e. terrain straight down) most?

Reply:
It doesn't start 10 pixels below, that's just the Vector, and it gets rotated for each ray

I understand it's weird that it starts with a downwards-facing vector instead of a forwards one though

Reply:
Yeah, it took me a while to figure out what was going on, but I think I get it.  Looking forwards would have been my first instinct, but looking downwards is probably smarter, cause I'd imagine it accounts for downwards slopes better.